docs/syntax_and_semantics/select.md

```diff
@@ -0,0 +1,52 @@
+# select
+
+The `select` expression chooses from a set of blocking operations and proceeds with the branch that becomes available first.
+
+## Syntax
+
+The expression starts with the keyword `select`, followed by a list of one or more `when` branches.
+Each branch has a condition and a body, separated by either
+a statement separator or the keyword `then`.
+Optionally, the last branch may be `else` (without condition).  This denotes the `select` action as non-blocking.
+The expression closes with an `end` keyword.
+
+> NOTE:
+> `select` is similar to a [`case` expression](./case.md) with all branches referring to potentially blocking operations.
+
+Each condition is either a call to a select action or an assignment whose right-hand side is a call to a select action.
+
+```crystal
+select
+when foo = foo_channel.receive
+  puts foo
+when bar = bar_channel.receive
+  puts bar
+when exit_channel.receive
+  exit
+when timeout(5.seconds)
+  puts "Timeout"
+end
+```
+
+## Select actions
+
+A select action call calls a method with the implicit suffix "_select_action".
```

Comment:
The `?` suffix means the transformation is `foo?` -> `foo_select_action?`, and `foo` -> `foo_select_action` 🕵️ .

I don't remember if `foo` and `foo_select_action` are constrained to have some relation in their type signature.

Reply:
I added a mention of the suffix.

> I don't remember if `foo` and `foo_select_action` are constrained to have some relation in their type signature.

I don't think so. There's no need for any constraints because *all* calls in a `select` expression's `where` condition are expanded. If a method signature doesn't match, the type checker will complain.

Reply:
They do.

```crystal
def foo_select_action : Channel::TimeoutAction
  timeout_select_action(1.second)
end

a =
  select
  when x = foo
    x
  else
    nil
  end
```

complains with

```
In foo.ign.cr:15:12

 15 | when x = foo
               ^--
Error: undefined local variable or method 'foo' for top-level
```

If we define `foo` with the wrong type

```
def foo : Int32
  0
end
```

we get

```
 14 | select
      ^
Error: can't cast (Channel::NotReady | Nil) to Int32
```

Since `Channel::TimeoutAction` includes `SelectAction(Nil)` we need `def foo : Nil` for things to compile

Reply:
Hm, interesting. I have no idea where in the compiler that would be happening, though 😆 
Literal expander just transforms the calls without any checking.